README.md
# CCML: A two-step consensus clustering inputing multiple predictive labels with different sample coverages (missing labels)

## Dependencies
------------
* diceR
* ggplot2
* parallel
* plyr
* tidyr
* SNFtools
* ConsensusClusterPlus



## Installation
------------
Get the latest released version from CRAN:    
``` r
install.packages("ccml")
```
Or from GitHub:

``` r
remotes::install_github("*****")
```
Or install locally:
``` r
install.packages("ccml_1.0.0.tar.gz",repos = NULL, type="source")
```

## Setup
------------
This package needs to create a folder to store the result. Let’s call the folder
`output` and assume the full path to this folder is
`xxxxx/output`. Run the function below to set the path for the package.

``` r
library(ccml)
setwd("xxxxx/output")
```

Or a folder named "output" will be created under the default path when the program is running.



## Introduction
------------
NULL



## Functions
------------
* ccml     
*A two-step consensus clustering inputing multiple predictive labels with different sample coverages (missing labels).*

* callNCW  
*Calculate normalized consensus weight(NCW) matrix based on permutation.*


* plotCompareCW  
*Plot of original consensus weights vs. normalized consensus weights grouping by the number of co-appeared percent of clustering(non-missing).*


* randConsensusMatrix   
*Calculate consensus weight matrix based on the permuted input label matrix. Internal function used by callNCW.*


* spectralClusteringAffinity  
*Perform spectral clustering algorithms for an affinity matrix, using SNFtool::spectralClustering.*




## Run examples
------------
In the repository, we give some examples to show how to run **ccml**.
*example_data is built into R package and can be loaded with data().*

* load data   
*A matrix or data frame of input labels or a character value of input file name, columns are different clustering results and rows are samples.*
  ``` r
  data(example_data)
  label=example_data
  ```

* output directory      
*This part is mentioned in Setup above.*
  ``` r
  title="output"
  ```
  
* run ccml
   ```r
    # not estimate stability of permutation numbers.
    res_1=ccml(title=title,label=label,nperm = 2,stability=F,maxK=3,pItem=0.8,plot = "pdf")

    # other methods for clustering of distance matrix
    res_2<-ccml(title=title,label=label,nperm = 10,stability=F,maxK=3,pItem=0.9,plot = "pdf",clusterAlg = "hc")

    # not output as "rdata"
    res_3<-ccml(title=title,label=label,nperm = 10,stability=F,maxK=3,pItem=0.9,plot = "pdf",output=FALSE)

    # The output is a list.Some examples are as follws
    # get consensusMatrix
    res$fcluster[[3]]$consensusMatrix

    # get consensusTree
    res$fcluster[[2]]$consensusTree

    # get consensusClass
    res$fcluster[[2]]$consensusClass

   ```

* run callNCW
   ```r
    ncw<-callNCW(title=title,label=label,nperm = 10, ncore = 1, seedn = 100, stability = TRUE)
   ```
   
* run plotCompareCW
   ```r
    plotCompareCW(title=title,label=label,ncw)
   ```
  

## Output Description
------------
* consensus     
  The heatmaps of the consensus matrices(for k= 2, 3, 4),consensus CDF Plot, Delta Area Plot and Tracking Plot are used to determine the best K value of clustering.
  

  <img src="img/k_2.jpg" width="29%">
  <img src="img/k_3.jpg" width="29%">
  <img src="img/k_4.jpg" width="29%">

  <img src="img/2.CDF.jpg" width="27%">
  <img src="img/2_Delta.jpg" width="27%">
  <img src="img/2_tracking.jpg" width="27%">


* icl  
  Item-Consensus Plot show the stability of members.   
  ![icl](img/3.icl.jpg)   
  ![icl_2](img/3.icl_2.jpg)

  
* plotCompareCW   
  This plot show the original consensus weights vs. normalized consensus weights grouping by the number of co-appeared percent of clustering(non-missing).The number of duplicates sample are indicated by the size of the colored portion, whose color corresponds to percent of co-appeared of clustering.   
  ![PCW](img/3.PCW.jpg)

* stability.seed100.n10K   
  The name of plot represents 10*1000 permutation using random numbers with seedn=100.The stability of normalized consensus weight is estimated based on permutation numbers.
  ![ncw](img/NCW.jpg)


NEWS.md
## version 1.2.0

---


### Complete ccml functions



## version 1.1.0

---


### Minor bugs fixed



## version 1.0.0

---

### This is the first version of the ccml package



R/ccml.R
ccml<-function(title, label, output="rdata",nperm = 10, ncore = 1, seedn = 100, stability = TRUE, maxK = 15, reps = 1000, pItem = 0.9, plot = "pdf", clusterAlg = "spectralClusteringAffinity", innerLinkage = "complete",...){
#' A two-step consensus clustering inputing multiple predictive labels with different sample coverages (missing labels)
#'
#' @param title A character value for output directory. Directory is created only if not existed. This title can be an abosulte or relative path. Input for \code{callNCW, plotCompareCW, ConsensusClusterPlus::ConsensusClusterPlus, ConsensusClusterPlus::calcICL}
#' @param label A matrix or data frame of input labels or a character value of input file name, columns=different clustering results and rows are samples. \code{label} could be import as '.rdata', '.rda', or '.csv'. Input for \code{callNCW, plotCompareCW}
#' @param output A character value for output format, or "rdata"(default) as save to .rdata, others will return to workspace.
#' @param nperm A integer value of the permutation numbers, or nperm=10(default), which means \code{nperm}*1000 times of permutation. Input for \code{callNCW}
#' @param ncore A integer value of cores to use, or ncore=1 (default). It's the input core numbers for the parallel computation in R package \code{parallel}. Input for \code{callNCW}
#' @param seedn A numerical value to set the start random seed for reproducible results, or seedn=100 (default). For every 1000 iteration, the seed will +1 to get repeat results. Input for \code{callNCW, ConsensusClusterPlus::ConsensusClusterPlus}
#' @param stability A logical value. Should estimate the stability of normalized consensus weight based on permutation numbers (default stability=TRUE), or not? Input for \code{callNCW}
#' @param maxK integer value. maximum cluster number to evaluate. Input for \code{ConsensusClusterPlus::ConsensusClusterPlus} for the consensus clustering based on normalized consensus weights.
#' @param reps integer value. number of subsamples. Input for \code{ConsensusClusterPlus::ConsensusClusterPlus}
#' @param pItem numerical value. proportion of items to sample. Input for \code{ConsensusClusterPlus::ConsensusClusterPlus}
#' @param plot character value. NULL - print to screen, 'pdf', 'png', 'pngBMP' for bitmap png, helpful for large datasets, or 'pdf'(default). Input for \code{ConsensusClusterPlus::ConsensusClusterPlus, ConsensusClusterPlus::calcICL}
#' @param clusterAlg character value. cluster algorithm. 'spectralClusteringAffinity' for spectral clustering of similarity/affinity matrix(default), other methods for clustering of distance matrix, 'hc' heirarchical (hclust), 'pam' for paritioning around medoids, 'km' for k-means upon data matrix, 'kmdist' for k-means upon distance matrices (former km option), or a function that returns a clustering. Input for \code{ConsensusClusterPlus::ConsensusClusterPlus}.
#' @param innerLinkage heirarchical linkage method for subsampling, or "complete"(default). Input for \code{ConsensusClusterPlus::ConsensusClusterPlus}
#' @param ... Other input arguments for \code{ConsensusClusterPlus::ConsensusClusterPlus}
#' @return A list of three items
#' \itemize{
#'   \item ncw - A matrix of normalized consensus weights. Output from \code{callNCW}.
#'   \item fcluster - A list of length maxK. Each element is a list containing consensusMatrix (numerical matrix), consensusTree (hclust), consensusClass (consensus class asssignments). ConsensusClusterPlus also produces images. Output from \code{ConsensusClusterPlus::ConsensusClusterPlus}
#'   \item icl a list of two elements clusterConsensus and itemConsensus corresponding to cluster-consensus and item-consensus. Output from \code{ConsensusClusterPlus::ConsensusClusterPlus}
#' }
#' @import ConsensusClusterPlus
#' @export
#' @examples
#'
#' ## load data
#' data(example_data)
#' label=example_data
#'
#' ## run ccml
#' ## "output" is a character value for output directory
#' #title="output"
#'
#' ## not estimate stability of permutation numbers.
#' #res_1=ccml(title=title,label=label,nperm = 2,stability=F,maxK=3,pItem=0.8,plot = "pdf")
#'
#' ## other methods for clustering of distance matrix
#' #res_2<-ccml(title=title,label=label,nperm = 10,stability=F,maxK=3,
#' #           pItem=0.9,plot = "pdf",clusterAlg = "hc")
#'
#' ## not output as "rdata"
#' #res_3<-ccml(title=title,label=label,output=FALSE,nperm = 10,stability=F,maxK=3,
#' #           pItem=0.9,plot = "pdf")

  requireNamespace("ConsensusClusterPlus")
  requireNamespace("tidyr")

  # if (!requireNamespace("BiocManager", quietly = TRUE))
  #   install.packages("BiocManager")

  # install_bioc("ConsensusClusterPlus")
  # usethis::use_dev_package("ConsensusClusterPlus")

  ## convert 'title' for output directory to formated absolute path
  ### add "/" at the end of title
  if (!substr(title,nchar(title),nchar(title))=="/") {
    title = paste0(title,"/")
  }

  if (substr(title,1,2)=="./") {
    ### relative path to absolute path
    title = paste0(getwd(),"/",substr(title,3,nchar(title)))
  } else if (substr(title,1,1)=="D" | substr(title,1,1)=="E" | substr(title,1,1)=="F"| substr(title,1,1)=="C"){
    ### character input to absolute path
    title = title
  }else if (!substr(title,1,1)=="/"){
    ### character input to absolute path
    title = paste0(getwd(),"/",title)
  }

  # ## import label
  # if (all(is.character(label)&grepl(".rdata$|.rda$",label))){
  #   load(label)
  # } else if (all(is.character(label)&grepl(".csv$",label))){
  #   label2 = read.csv2(label,header = T)
  #   label = label2[,2:ncol(label2)]
  #   rownames(label) = label2[,1]
  #   rm(label2)}


  # calculate normalized consensus weight matrix
  message("Main Step 1 - Calculating normalized consensus weight matrix")
  ncw <- callNCW(title, label, nperm = nperm, ncore = ncore, seedn = seedn, stability = TRUE)
  # plot of comparison of original and normalized consensus weights
  message("Main Step 2 - Comparison between original and normalized consensus weights")
  plotCompareCW(title, label, ncw)

  # The second step consensus clustering

  message("Main Step 3 - The 2nd step of consensus clustering based on normalized consensus weights")
  fcluster = ConsensusClusterPlus(as.dist(ncw), maxK=maxK, reps=reps, pItem=pItem, title=title, plot=plot, clusterAlg=clusterAlg, seed=seedn,innerLinkage=innerLinkage,...)
  # network as an affinity matrix can be used only because the "spectalClustering" algorithm use affinity matrices; keep in mind that classic algorithms use distance matrices
  icl = calcICL(fcluster, title=title, plot=plot)
  res <- list(ncw=ncw,fcluster=fcluster,icl=icl)
  if (output=="rdata"){
    save(res,file=paste0(title,"s",seedn,"n",nperm,"res.rdata"))
    res = "Output to .rdata"
  }
  return(res)
}


R/plotCompareCW.R
plotCompareCW<- function(title,label,ncw){
  #' Plot of original consensus weights vs. normalized consensus weights grouping by the number of co-appeared percent of clustering(non-missing).
  #'
  #' @param title A character value for output directory.
  #' @param label A matrix or data frame of input labels, columns=different clustering results and rows are samples.
  #' @param ncw A matrix of normalized consensus weights with sample-by-sample as the order of sample(rows) in \code{label}.
  #' @return A ggplot point in PDF format with x-axis: original consensus weights; y-axis: normalized consensus weights; color: percent of co-appeared of clustering; size: number of duplicates sample .
  #' @import diceR  ggplot2 plyr
  #' @export
  #' @examples
  #'
  #' ## load data
  #' data(example_data)
  #' label=example_data
  #'
  #' ## "output" is a character value for output directory
  #' #title="output"
  #'
  #' #ncw<-callNCW(title=title,label=label)
  #' #plotCompareCW(title=title,label=label,ncw)




  requireNamespace("diceR")
  requireNamespace("ggplot2")
  requireNamespace("tidyr")
  requireNamespace("plyr")

  # require(diceR);require(ggplot2);require(plyr)

  ## convert 'title' for output directory to formated absolute path
  ### add "/" at the end of title
  if (!substr(title,nchar(title),nchar(title))=="/") {
    title = paste0(title,"/")
  }

  if (substr(title,1,2)=="./") {
    ### relative path to absolute path
    title = paste0(getwd(),"/",substr(title,3,nchar(title)))
  }else if (substr(title,1,1)=="D"){
    ### character input to absolute path
    title = title
  }else if (!substr(title,1,1)=="/"){
    ### character input to absolute path
    title = paste0(getwd(),"/",title)
  }

  ### original consensus weight matrix of label
  cw <- diceR::consensus_matrix(label)

  ### pair wise index of sample pairs with non missing values
  pair.ind = which(!is.na(cw),arr.ind = T)
  pair.ind = pair.ind[pair.ind[,1]-pair.ind[,2]<0,]

  ### number of shared non missing clusterings for each pair of samples
  nclust <- sapply(1:nrow(pair.ind),function(a){sum(colSums(is.na(label[pair.ind[a,],]))==0)})/ncol(label)
  nclust = cut(nclust, breaks = seq(0,1,by=0.1),include.lowest = T) # convert to percentage

  # data frame for plot
  pdata <- cbind.data.frame(cw=cw[pair.ind],ncw=ncw[pair.ind],nclust=nclust)
  pdata=count(pdata,vars=colnames(pdata))
  pdata$freq = cut(pdata$freq, breaks = unique(floor(quantile(pdata$freq,probs=seq(0,1,0.1)))),include.lowest = T,ordered_result = T,dig.lab=ceiling(log10(max(pdata$freq)))+1) # cut frequency to 10 breaks

  colnames(pdata)[3] = "Clustering"
  colnames(pdata)[4] = "Duplicates"

  ggplot(data=pdata,aes(x=cw,y=ncw,color=Clustering,size=Duplicates)) +
    geom_point() +
    scale_colour_brewer(palette="Spectral",name="Clustering") +
    xlab("Original consensus weights") +
    ylab("Normalized consensus weights") +
    theme_bw(base_size = 14)

  ggsave(paste0(title,"plotCompareCW.pdf"),width=8,height=6)
}


R/randConsensusMatrix.R
randConsensusMatrix <- function(l.seed,l.label=label,l.ns=ns,l.nc=nc,l.nv=nv,l.index=index,l.pair.ind=pair.ind,l.ppath=ppath){
  #' Calculate consensus weight matrix based on the permuted input label matrix. Internal function used by \code{callNCW}
  #' @param l.seed A numerical value to set the random seed for reproducible results, 1000 random label matrix will be generated based on this seed number.
  #' @param l.label A matrix or data frame of input labels, columns=different clustering results and rows are samples.
  #' @param l.ns A integer value of number of samples, =\code{nrow(l.label)}
  #' @param l.nc A integer value of number of samples, =\code{ncol(l.label)}
  #' @param l.nv A integer vector of the number of non missing values for each column in \code{l.label}
  #' @param l.index A list of index with length of \code{l.nc} of non missing values for each column in \code{l.label}
  #' @param l.pair.ind A n-by-2 index matrix of array indices of upper triangular of \code{l.label} with non missing values
  #' @param l.ppath A character value for output directory.
  #' @return A character of finished seed.
  #' @return Write a binary file of 1000 random consensus weight matrix(as a vector n-by-1, n= nrow(\code{l.pair.ind})) with the seed \code{l.seed}, output file name: paste0("s",\code{l.seed},"rcw").
  #' @import diceR
  #' @export
  #'
  #'


  # require(diceR)
  requireNamespace("diceR")

  fname = paste0(l.ppath,paste0("s",l.seed,"rcw"))
  cc.rand <- lapply(1:1000,function(b){

    tmp.sample <- lapply(l.nv,sample);
    tmp.group <- sapply(1:l.nc,function(i){
      tmp.rand <- rep(NA,length=l.ns)
      tmp.rand[l.index[[i]]] <- l.label[l.index[[i]],i][tmp.sample[[i]]]
      return(tmp.rand)
    });
    return(tmp.group)})
  res <- lapply(cc.rand,diceR::consensus_matrix)
  res <- sapply(res,function(a){a[l.pair.ind]})
  t = as.vector(res)
  writeBin(t,fname)

  return(paste0("Finished", l.seed))

}


R/callNCW.R
callNCW <- function(title,label,nperm = 10, ncore=1,seedn=100,stability=TRUE) {
  #' Calculate normalized consensus weight(NCW) matrix based on permutation.
  #'
  #' @param title A character value for output directory. Directory is created only if not existed. This title can be an abosulte or relative path.
  #' @param label A matrix or data frame of input labels, columns=different clustering results and rows are samples.
  #' @param nperm A integer value of the permutation numbers, or nperm=10(default), which means \code{nperm}*1000 times of permutation.
  #' @param ncore A integer value of cores to use, or ncore=1 (default). It's the input core numbers for the parallel computation in R package \code{parallel}.
  #' @param seedn A numerical value to set the start random seed for reproducible results, or seedn=100 (default). For every 1000 iteration, the seed will +1 to get repeat results.
  #' @param stability A logical value. Should estimate the stability of normalized consensus weight based on permutation numbers (default stability=TRUE), or not?
  #' @export
  #' @import diceR parallel tidyr ggplot2
  #' @return A matrix of normalized consensus weights.
  #' @examples
  #'
  #' ## load data
  #' data(example_data)
  #' label=example_data
  #'
  #' ## "example_output" is a character value for output directory
  #' #title="output"
  #'
  #' ## run ncw
  #' #ncw<-callNCW(title=title,label=label)
  #'
  #'




  ## checking input----
  message("Step 1/3: Checking of inputs...")

  # message(paste0("shuru!!  title:",title))


  requireNamespace("diceR")
  requireNamespace("parallel")
  requireNamespace("tidyr")
  requireNamespace("ggplot2")
  requireNamespace("ConsensusClusterPlus")


  # require(diceR); require(parallel); require(tidyr);require(ggplot2)
  if(missing(title))
  { stop("A 'title' of a character value for output directory must be provided") }
  if(!is.character(title))
  { stop("A 'title' of a character value for output directory must be provided") }
  if(missing(label))
  { stop("A input 'label' as matrix or data frame must be provided") }
  if(!is.matrix(label)&!is.data.frame(label))
  { stop("A input 'label' as matrix or data frame must be provided") }

  ## convert 'title' for output directory to formated absolute path
  ### add "/" at the end of title
  if (!substr(title,nchar(title),nchar(title))=="/") {
    title = paste0(title,"/")
  }

  if (substr(title,1,2)=="./") {
    ### relative path to absolute path
    message(paste0("The first choice:",title))
    title = paste0(getwd(),"/",substr(title,3,nchar(title)))
    # message(paste0("The first choice:",title))
  } else if (substr(title,1,1)=="D" | substr(title,1,1)=="E" | substr(title,1,1)=="F"| substr(title,1,1)=="C"){
    ### character input to absolute path
    title = title
  }else if (!substr(title,1,1)=="/"){
    ### character input to absolute path
    title = paste0(getwd(),"/",title)
    # message(paste0("The second choice:",title))
  }

  ## path for cache permuation result
  ppath = paste0(title,"permutation/")

  ## create folders of title and title/permutation ----
  if (!dir.exists(title)){
    dir.create(title,recursive = T)
    message(paste("- Created a new folder:\n", title))
    dir.create(ppath,recursive = T)
  } else {
    message(paste(" - The directory already existed.\nNew results will be added to:\n", title))
    if (!dir.exists(ppath)){
      dir.create(ppath,recursive = T)
      message(paste(" - Created a new folder for cache permutation results:\n", ppath))} else {
        message(paste(" - The cache directory already existed.\n -New permutation results will be added to:\n", ppath))
      }
  }

  ## permute labels------
  step = 1000 # number of random consensus weight matrix in each seed

  ## mc.cores, number of cores will be used
  ncore = min(parallel::detectCores(),ncore)


  ### number of non missing values for each column
  nv <- colSums(!is.na(label))
  nc <- ncol(label) # number of clustering labels
  ns <- nrow(label) # number of samples
  index <- apply(label,2,function(b){which(!is.na(b))})
  ### list of index of non missing values for each column


  ### consensus matrix of label
  cm <- diceR::consensus_matrix(label)

  ### pair wise index of sample pairs with non missing values
  pair.ind = which(!is.na(cm),arr.ind = T)
  pair.ind = pair.ind[pair.ind[,1]-pair.ind[,2]<0,]

  ### check existed randome consensus weight matrix in the results folder
  filename = dir(ppath)
  filename = filename[grepl("rcw$",filename)]    ##find "rcw$" in ppath
  if (length(filename)>0) {
    filename = gsub("s","",filename)
    seede = as.integer(gsub("rcw","",filename)) # existed seed results

  } else {seede=NULL}
  seeda = seedn+0:(nperm-1) # all seeds
  seedL = setdiff(seeda,seede) # seeds need to be run

  ## calculate rand consensus matrix, output to binary files----


  message(paste("\nStep 2/3: Calculating of", nperm*step, "random consensus matrixes...\n"))
  ppath = paste0(title,"permutation/")


  randL <- parallel::mclapply(seedL,function(x){randConsensusMatrix(l.seed=x,l.label=label,l.ns=ns,l.nc=nc,l.nv=nv,l.index=index,l.pair.ind=pair.ind,l.ppath=ppath)},mc.cores=ncore)

  ## calculate NCW----
  ### import all random consensus matrixes to randCM


  randCM<- lapply(seeda,function(a){
    fname = file.path(ppath,paste0("s",a,"rcw"))
    tmp <- readBin(fname,"numeric",n=nrow(pair.ind)*step)
    tmp = matrix(tmp,nrow=nrow(pair.ind),ncol=step)
    return(tmp)
  })

  randCM <- as.matrix(cbind.data.frame(randCM))


  ## estimation of stability of permutation number----
  if (stability==TRUE) {
    # calculate normalized censensus weight for every step(1000)
    message(paste("Step3/3: Calculating & Estimating the stability of permuation number for normalized consensus weight...\n"))
    res.ecdf <- sapply(1:nrow(pair.ind),function(a){
      v = cm[pair.ind[a,1],pair.ind[a,2]]
      t = randCM[a,]
      r = sapply(seq(step,ncol(randCM),length.out=length(seeda)),function(b){ecdf(t[1:b])(v)})
      return(r)
    })
    res.ecdf <- t(res.ecdf)
    colnames(res.ecdf) <- seq(step,ncol(randCM),length.out=length(seeda))

    ## for 0 and 1 setup min and max values based on permutation numbers
    for (i in 1:ncol(res.ecdf)){
      res.ecdf[res.ecdf[,i]==0,i] = 1/(1000*i)
      res.ecdf[res.ecdf[,i]==1,i] = 1-1/(1000*i)
    }

    res.ncw <- res.ecdf[,ncol(res.ecdf)]
    # the difference (Squared Euclidean Distance of NCW) for every step(1000) permutation
    res.ecdf.diff <- matrix(0,nrow=nrow(res.ecdf),ncol=ncol(res.ecdf)-1)
    for(i in 2:ncol(res.ecdf)){
      res.ecdf.diff[,i-1]=(res.ecdf[,i]-res.ecdf[,(i-1)])^2

    }

    colnames(res.ecdf.diff) = colnames(res.ecdf)[2:ncol(res.ecdf)]




    ### prepare pdata for plot, output to paste0(title,".seed",seedn,".n",nperm,"K.pdf")
    ### squared euclidean distance for each cell
    a<-as.data.frame(cbind.data.frame(pair=1:nrow(res.ecdf.diff),
                                         res.ecdf.diff))
    b<-tidyr::gather(a,colnames(res.ecdf.diff),key="reps",value= "diff2")
    b<-as.data.frame(b)
    pdata<-b


    class(pdata$reps) = "integer"
    pdata$reps = pdata$reps/step
    class(pdata$reps) = "integer"

    ### overall squared euclidean distance
    pdata2 <- cbind.data.frame(reps=as.integer(colnames(res.ecdf.diff)),y=colSums(res.ecdf.diff))

    pdata2$reps = pdata2$reps/step
    class(pdata2$reps) = "integer"

    sf = max(pdata2$y)/max(pdata$diff2) # scaling factor
    ggplot2::ggplot() +
      geom_boxplot(pdata, mapping=aes(x=reps, y=diff2*sf,group=reps)) +
      geom_line(data=pdata2,aes(x=reps, y=y,group=1),color="blue") +
      geom_point(data=pdata2,aes(x=reps, y=y,group=1),size=3,shape=23,fill = "blue") +
      scale_y_continuous(name="Overall Squared Euclidean Distance of NCW",sec.axis = sec_axis(~ ./sf, name="Squared Euclidean Distance of NCW"))+
      scale_x_continuous(name="Permutation numbers(*1000)",breaks=unique(pdata2$reps),labels=as.character(unique(pdata2$reps)))+
      ggtitle("Stability of permutation numbers for NCW calculation") +
      theme_bw(base_size = 14) +
      theme(axis.text.y.left = element_text(color = "blue"),plot.title = element_text(hjust = 0.5))

    ggplot2::ggsave(paste0(title,"stability.seed",seedn,".n",nperm,"K.pdf"),width=max(8,nperm*0.5),height=5)
    message(paste("Finished estimation of stability of permutation numbers for normalized consensus weight:\n"))

  } else {
    # calculate normalized censensus weight for all permutated results
    message(paste("Step3/3: Calculating the matrix of normalized consensus weight...\n"))
    res.ecdf <- sapply(1:nrow(pair.ind),function(a){
      v = cm[pair.ind[a,1],pair.ind[a,2]]
      t = randCM[a,]
      r = ecdf(t)(v)
      return(r)
    })
    res.ncw <- res.ecdf

  }


  res.ncw[res.ncw==1] = 1-1/(1000*nperm) # max based on permutation times
  res.ncw[res.ncw==0] = 1/(1000*nperm) # min


  ## output to matrix of NCW ----
  ncw <- cm
  ncw[pair.ind]=res.ncw
  ncw[pair.ind[,c(2,1)]]=res.ncw

  ### add rownames to ncw
  if (!is.null(rownames(label))) {
    rownames(ncw) = rownames(label)
    colnames(ncw) = rownames(label)
  }

  return(ncw)
}
